File: restdoc-rpc-client/restdoc-client-api/src/main/kotlin/restdoc/client/api/model/InvocationResult.java
package restdoc.client.api.model;

import restdoc.remoting.protocol.RemotingSerializable;


/**
 * InvocationResult
 *
 * @author Maple
 */
public class InvocationResult extends RemotingSerializable {

    private Boolean isSuccessful = true;
    private String exceptionMsg;

    private Invocation invocation;


    public InvocationResult() {
    }


    public InvocationResult(Boolean isSuccessful, String exceptionMsg, Invocation invocation) {
        this.isSuccessful = isSuccessful;
        this.exceptionMsg = exceptionMsg;
        this.invocation = invocation;
    }

    public Boolean getSuccessful() {
        return isSuccessful;
    }

    public void setSuccessful(Boolean successful) {
        isSuccessful = successful;
    }

    public String getExceptionMsg() {
        return exceptionMsg;
    }

    public void setExceptionMsg(String exceptionMsg) {
        this.exceptionMsg = exceptionMsg;
    }

    public Invocation getInvocation() {
        return invocation;
    }

    public void setInvocation(Invocation invocation) {
        this.invocation = invocation;
    }
}

File: restdoc-rpc-client/restdoc-client-api/src/main/kotlin/restdoc/client/api/model/InvocationResult.kt
package restdoc.client.api.model

import restdoc.remoting.protocol.RemotingSerializable

/**
 * InvocationResult
 */
open class InvocationResult : RemotingSerializable {
    var isSuccessful: Boolean = true
    var exceptionMsg: String? = null


    lateinit var invocation: Invocation

    constructor()
    constructor(isSuccessful: Boolean, exceptionMsg: String?, invocation: Invocation) {
        this.isSuccessful = isSuccessful
        this.exceptionMsg = exceptionMsg
        this.invocation = invocation
    }
}

/**
 * T must be serilizable
 */
class DubboInvocationResult : InvocationResult {

    var returnValue: String? = ""
    var returnValueType: String = Void::class.java.name

    constructor()
    constructor(isSuccessful: Boolean, exceptionMsg: String?,
                invocation: Invocation,
                returnValueType: String,
                returnValue: String?
    ) : super(isSuccessful, exceptionMsg, invocation) {
        this.returnValue = returnValue
        this.returnValueType = returnValueType
    }
}


/**
 * RestWebInvocationResult
 */
class RestWebInvocationResult : InvocationResult {
    var status: Int = 200
    var responseHeaders: MutableMap<String, MutableList<String>> = mutableMapOf()
    var responseBody: Any? = null

    constructor()
    constructor(isSuccessful: Boolean,
                exceptionMsg: String?,
                invocation: Invocation,
                status: Int,
                responseHeaders: MutableMap<String, MutableList<String>>,
                responseBody: Any?
    ) : super(isSuccessful, exceptionMsg, invocation) {
        this.status = status
        this.responseHeaders = responseHeaders
        this.responseBody = responseBody
    }
}